### PR TITLE
[Feature][MRV2] Support MTP with DSA in eager mode

### PR DESCRIPTION
**Pull request by jiajinzhu2**
### What this PR does / why we need it?

This PR fixes DeepSeek-V4 MTP inference with ModelRunner V2 on the Ascend DSA attention path.

Two issues prevented MTP from running:

- `DeepSeekMultiTokenPredictorLayer` returned hidden states with shape `[num_tokens, hc_mult, hidden_size]`, while the next MTP step expects `[num_tokens, hc_mult * hidden_size]`.
- The V2 autoregressive speculator incorrectly treated the draft attention as MLA based on the target model configuration. It therefore applied the generic MLA/GQA metadata initialization and update logic to `AscendDSAMetadata`, including access to the unavailable `seq_lens_cpu` field. DSA manages its per-step slot mapping, SAS, QLI, and sequence-length state through its own metadata builder and should bypass these generic updates.

The patch flattens the reusable DSV4 MTP hidden states and updates sequence lengths through `seq_lens_cpu` when available, falling back to the DSA device tensor otherwise. Existing metadata types and the ModelRunner V1 path are unchanged.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?
```
export VLLM_VERSION=0.26.1
export HCCL_OP_EXPANSION_MODE="AIV"
export HCCL_BUFFSIZE=1024
export PYTORCH_NPU_ALLOC_CONF="expandable_segments:True"
export OMP_PROC_BIND=false
export OMP_NUM_THREADS=1
export TASK_QUEUE_ENABLE=1
export ASCEND_RT_VISIBLE_DEVICES="0,1,2,3,4,5,6,7"
vllm_cmd=(
  "$vllm_bin" serve /mnt/a800_weight/DeepSeek-V4-Flash-w8a8-mtp/
  --host 0.0.0.0
  --port 2723
  --served-model-name model

  --data-parallel-size 1
  --tensor-parallel-size 8
  --pipeline-parallel-size 1
  --enable-expert-parallel
  --prefill-context-parallel-size 1
  --decode-context-parallel-size 1

  --max-model-len 4096
  --max-num-seqs 8
  --max-num-batched-tokens 512
  --gpu-memory-utilization 0.8

  --enforce-eager
  --no-enable-prefix-caching
  --quantization ascend
  --safetensors-load-strategy prefetch
  --additional-config '{
    "enable_dsa_cp": false,
    "enable_fused_mc2": 0,
    "enable_flashcomm1": false,
    "enable_balance_scheduling": false
  }'
)
```
ModelrunnerV1
```
{
  "num_drafts": 48672.0,
  "num_accepted_tokens_per_pos": [
    45109.0,
    34383.0,
    17160.0
  ],
  "acceptance_per_pos": [
    0.9267956936226167,
    0.70642258382643,
    0.3525641025641026
  ],
  "acceptance_len": 2.9857823800131493
}

```
ModelrunnerV2
```
{
  "num_drafts": 48505.0,
  "num_accepted_tokens_per_pos": [
    44887.0,
    34117.0,
    17074.0
  ],
  "acceptance_per_pos": [
    0.9254097515720029,
    0.7033707865168539,
    0.352004947943511
  ],
  "acceptance_len": 2.980785486032368
}
```

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py [hunk 1]
```diff
@@ -486,7 +490,10 @@ def _update_decode_attn_metadata(
             )  # .decode (MLA) / top-level (GQA)
             decode_metadata.seq_lens_list = seq_lens_list
             decode_metadata.actual_seq_lengths_q = query_lens_list
-            metadata.seq_lens_cpu.copy_(next_seq_lens_cpu)
+            target_seq_lens = getattr(metadata, "seq_lens_cpu", None)
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py]: Judge it with `self.attn_architecture`.
Don't use `getattr`.

File: vllm_ascend/models/deepseek_v4_mtp.py [hunk 2]
```diff
@@ -130,7 +130,8 @@ def forward(
         # hidden_states = self.hc_head(hidden_states, self.hc_head_fn,
         #                              self.hc_head_scale, self.hc_head_base)
 
-        return hidden_states
+        # Reused as previous_hidden_states by the next speculative step.
```
Comment (drslark) [inline vllm_ascend/models/deepseek_v4_mtp.py]: What is it for?

**Review by drslark — CHANGES_REQUESTED** [review]

File: vllm_ascend/models/deepseek_v4_mtp.py (hunk 2 above)
Comment (jiajinzhu2) [reply vllm_ascend/models/deepseek_v4_mtp.py]: `DeepSeekMultiTokenPredictorLayer` returned `hidden states` with shape `[num_tokens, hc_mult, hidden_size]`, while the next MTP step expects `[num_tokens, hc_mult * hidden_size]`.

**Comment by drslark** [comment]
Add an `e2e`, please.

**Comment by wxsIcey** [comment]
We can prevent the DSA-type model from updating `seq_lens_cpu` on the CPU side.

**Review by lilinsiman — APPROVED** [review]

File: vllm_ascend/models/deepseek_v4_mtp.py (hunk 2 above)
Comment (jiajinzhu2) [reply vllm_ascend/models/deepseek_v4_mtp.py]: `flatten(1)` adapts DeepSeek V4 MTP’s internal `[num_tokens, hc_mult, hidden_size]` output to MRV2’s `[num_tokens, hc_mult * hidden_size]` hidden-state buffer contract. It does not change V1 results because V1 already performs the equivalent `view(batch_size, -1)` before reusing the hidden states in the next draft step. `compute_logits()` also restores either representation to `[num_tokens, hc_mult, hidden_size]` before `hc_head`, so the logits remain unchanged.

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py [hunk 3]
```diff
@@ -469,7 +477,11 @@ def _update_decode_attn_metadata(
         if attn_metadata is None:
             return
 
-        num_reqs_padded = next(iter(attn_metadata.values())).seq_lens_cpu.shape[0]
+        attn_meta = next(iter(attn_metadata.values()))
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py]: Good.

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py [hunk 4]
```diff
@@ -421,10 +424,17 @@ def _init_decode_draft_attn_metadatas(self, attn_metadata: dict[str, Any] | None
         if attn_metadata is None:
             return
 
+        attn_backend = list(self.attn_backends.values())[0]
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py]: Move to `set_attn`.
Use to `list(self.attn_backends.values())[0]` generate `self.attn_architecture`.
Or make it iterate `attn_backend`s of layers.

**Review by drslark — CHANGES_REQUESTED** [review]

**Comment by jiajinzhu2** [comment]
> We can prevent the DSA-type model from updating `seq_lens_cpu` on the CPU side.

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py [hunk 5]
```diff
@@ -207,6 +209,13 @@ def set_attn(
                 attn_backends[layer_name] = attn_backend
 
         self.attn_backends = attn_backends
+        first_attn_backend = list(self.attn_backends.values())[0]
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py:212]: Good.

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py [hunk 6]
```diff
@@ -207,6 +209,13 @@ def set_attn(
...
+        if issubclass(first_attn_backend, AscendDSABackend):
+            self.attn_architecture = "DSA"
+        elif issubclass(first_attn_backend, AscendMLABackend):
+            self.attn_architecture = "MLA"
+        else:
+            self.attn_architecture = None
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py]: Add `GQA`, otherwise maybe raise?

**Review by drslark — CHANGES_REQUESTED** [review]

File: vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py (hunk 6 above)
Comment (jiajinzhu2) [reply vllm_ascend/worker/v2/spec_decode/autoregressive/speculator.py]: Ok, I have changed it!

**Review by drslark — APPROVED** [review]
LGTM.

**Review by Tflowers-0129 — APPROVED** [review]

**Review by linfeng-yuan — APPROVED** [review]
